NiKomLib/Terminal.c

```diff
@@ -156,3 +156,220 @@ void __saveds __asm LIBStripAnsiSequences(register __a0 char *ansistr, register
+*               ISO-8859-1 are replaced by '?'. For CHRS_CP437 and
+*               CHRS_MAC8 only characters greater than 128 are
+*               converted. For CHRS_SIS7 only characters greater than
+*               32 are converted.  Character greater than 128 in
+*               SIS-strings are interpreted as ISO-8859-1.
+*               CHRS_LATIN1 is not converted.
+*
+*               Note that the destination string is not nul terminated
+*               even if len is passed as zero. That must be done by
+*               the caller if needed.
+*/
+
+int __saveds __asm LIBConvMBChrsToAmiga(register __a0 char *dst,
+                                        register __a1 char *src,
+                                        register __d0 int len,
+                                        register __d1 int chrs,
+                                        register __a6 struct NiKomBase *NiKomBase) {
+  int si, di;
+
+  if(!len) {
+    len = INT_MAX;
+  }
+  for(si=0,di=0; src[si] && si<len; si++,di++) {
+    switch(chrs) {
+    case CHRS_CP437:
+      if(src[si] >= 128) {
+        dst[di] = NiKomBase->IbmToAmiga[src[si]];
+        continue;
+      }
+      /* No conversion, just copy. */
+      break;
+    case CHRS_SIS7:
+      if(src[si] >= 32) {
+        dst[di] = NiKomBase->SF7ToAmiga[src[si]];
+        continue;
+      }
+      /* No conversion, just copy. */
+      break;
+    case CHRS_MAC:
+      if(src[si] >= 128) {
+        dst[di] = NiKomBase->MacToAmiga[src[si]];
+        continue;
+      }
+      /* No conversion, just copy. */
+      break;
+    case CHRS_CP850:
+      dst[di] = convnokludge(src[si]);
+      continue;
+    case CHRS_UTF8:
+      if((src[si] & 0x80)) {
+        unsigned chr = (unsigned char)src[si];
+        int need;
+
+        if((chr & 0xc0) != 0xc0) {
+          /* Unexpected continuation byte. */
+          dst[di] = '?';
+          continue;
+        }
+        need = utf8_extra[chr - 192];
+        if(len == INT_MAX) {
+          unsigned idx = si+1;
+          switch(need) {
+          case 5:
+            if(src[idx] == '\0') {
+              return -5;
+            }
+            ++idx;
+          case 4:
+            if(src[idx] == '\0') {
+              return -4;
+            }
+            ++idx;
+          case 3:
+            if(src[idx] == '\0') {
+              return -3;
+            }
+            ++idx;
+          case 2:
+            if(src[idx] == '\0') {
+              return -2;
+            }
+            ++idx;
+          case 1:
+            if(src[idx] == '\0') {
+              return -1;
+            }
+            break;
+          }
+        } else if(need + 1 > (len - si)) {
+          return -(need + 1 - (len - si));
+        }
+        switch (need) {
+        case 1:
+          chr &= 0x1f;
+          chr <<= 6;
+          si++;
+          chr |= src[si] & 0x3f;
+          if (chr < 256) {
+            dst[di] = chr;
+          } else {
+            /* Non ISO-8859-1 */
```

Comment:
Non ANSI-C comments are ok so you can put `// Non ISO-8859-1` after the code on the same line.